### DISCUSSION TITLE
read hashed id programmatically in app router

### DISCUSSION THREAD

**Discussion by cryptod00c** [vercel/next.js#50770] [App Router]
how can I access the hashed id programmatically in the new router?

https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#scrolling-to-an-id

basically programmatically want to be able to access the `scrolling-to-an-id` so I can highlight it.

this is in a `use client` component, so I should be able to access all the router apis

**Comment by cryptod00c** [comment]
https://github.com/dylanmeivis/nextjs13-starter-markdoc/blob/83ac6899911f39733938db014db479497ff268f4/src/app/blog/(blog.components)/table.of.contents.tsx#LL11C39-L11C39

this will throw a server mismatch error

**Comment by icyJoseph** [comment]
Since you can't really access the `#` fragment part of the URL, you are necessarily, bound to read this client side. To make matters slightly worse, you also need to render things Server Side, to keep up with SEO and what not...

> The # fragment of a URL is never, ever, sent by the browser, to the server: https://stackoverflow.com/a/940996

I guess we are doomed? Well not just yet! **_CSS_** to the rescue: `:target`

- https://developer.mozilla.org/en-US/docs/Web/CSS/:target
- Target selector in TailWind: https://tailwindcss.com/docs/hover-focus-and-other-states#target

For actual programmatic usage is a bit tricky, but basically on every user action/browser event you'd just read the `window.location.hash` item, and for every time yo need to listen to the actual router changes, you need to use `useParams()` - https://github.com/vercel/next.js/discussions/49465 - However, for your use case, CSS' `:target` is the way to go.
